cookbooks/planet/files/default/replication-bin/replicate-changesets

```diff
@@ -207,63 +277,21 @@ class Replicator
     @config["data_dir"] + format("/%03d/%03d/%03d", sequence / 1000000, (sequence / 1000) % 1000, (sequence % 1000))
   end
 
-  def write_tmp_files!(changesets)
-    data_file = data_stem + ".osm.gz"
-    tmp_state = @config["state_file"] + ".tmp"
-    tmp_data = data_file + ".tmp"
-
-    FileUtils.mkdir_p(File.dirname(data_file))
-    Zlib::GzipWriter.open(tmp_data) do |fh|
-      fh.write(changeset_dump(changesets))
-    end
-    File.open(tmp_state, "w") do |fh|
-      fh.write(YAML.dump(@state))
-    end
-
-    # fsync the files in their old locations.
-    fsync(tmp_data)
-    fsync(tmp_state)
-
-    # sync the directory as well, to ensure that the file is reachable
-    # from the dirent and has been updated to account for any allocations.
-    fdirsync(File.dirname(tmp_data))
-    fdirsync(File.dirname(tmp_state))
-
-    # sanity check: the files we're moving into place
-    # should be non-empty.
-    raise "Temporary gzip file should exist, but doesn't." unless File.exist?(tmp_data)
-    raise "Temporary state file should exist, but doesn't." unless File.exist?(tmp_state)
-    raise "Temporary gzip file should be non-empty, but isn't." if File.zero?(tmp_data)
-    raise "Temporary state file should be non-empty, but isn't." if File.zero?(tmp_state)
-  end
-
-  def move_tmp_files_into_place!
-    data_file = data_stem + ".osm.gz"
-    data_state_file = data_stem + ".state.txt"
-    tmp_state = @config["state_file"] + ".tmp"
-    tmp_data = data_file + ".tmp"
-
-    FileUtils.mv(tmp_data, data_file)
-    FileUtils.cp(tmp_state, @config["state_file"])
-    FileUtils.mv(tmp_state, data_state_file)
-
-    # fsync the files in their new locations, in case the inodes have
-    # changed in the move / copy.
-    fsync(data_file)
-    fsync(@config["state_file"])
-    fsync(data_state_file)
-
-    # sync the directory as well, to ensure that the file is reachable
-    # from the dirent and has been updated to account for any allocations.
-    fdirsync(File.dirname(data_file))
-    fdirsync(File.dirname(@config["state_file"]))
-  end
-
   # saves new state (including the changeset dump xml)
   def save!
-    File.open(@config["state_file"], "r") do |fl|
+    state_file = @config["state_file"]
+    lock_file = @config["lock_file"]
+
+    File.open(lock_file, File::RDWR | File::CREAT) do |fl|
       fl.flock(File::LOCK_EX)
 
+      @state = read_state(state_file)
+
+      # initialise this stuff _after_ we've got the lock on the state file.
+      @conn = PG::Connection.connect(@config["db"])
+      # get current time from the database rather than the current system
+      @now = @conn.exec("select now() as now").map { |row| Time.parse(row["now"]) }[0]
```

Comment:
If _making replication more robust_ includes dealing with a non-utc server tz, something like `now() at time zone 'utc'`  might be helpful here.